File: vermyth/engine/policy/rule_based.py
from __future__ import annotations

from vermyth.engine.policy.base import PolicyModel
from vermyth.schema import (
    DivergenceStatus,
    PolicyAction,
    PolicyThresholds,
    ScoreComponent,
    VerdictType,
)


class RuleBasedPolicyModel(PolicyModel):
    name = "rule_based"
    version = "1"

    def decide(
        self,
        *,
        verdict: VerdictType,
        adjusted_resonance: float,
        divergence_status: DivergenceStatus | None,
        narrative_coherence: float | None,
        scores: list[ScoreComponent],
        aggregate_score: float,
        thresholds: PolicyThresholds,
    ) -> tuple[PolicyAction, str]:
        drift_ok = divergence_status is None or divergence_status in (
            DivergenceStatus.STABLE,
            thresholds.max_drift_status,
        )
        effect_sc = next((c for c in scores if c.name == "effect_risk"), None)
        if (
            thresholds.effect_risk_min_score is not None
            and effect_sc is not None
            and effect_sc.weight > 0
            and float(effect_sc.value) < float(thresholds.effect_risk_min_score)
        ):
            return (
                PolicyAction.DENY,
                f"verdict={verdict.value}; effect_risk={effect_sc.value:.3f} below min "
                f"{thresholds.effect_risk_min_score:.3f}; {effect_sc.explanation}",
            )
        if verdict == VerdictType.INCOHERENT or divergence_status == DivergenceStatus.DIVERGED:
            action = PolicyAction.DENY
        elif (
            verdict == VerdictType.COHERENT
            and adjusted_resonance >= float(thresholds.allow_min_resonance)
            and drift_ok
            and aggregate_score >= float(thresholds.allow_min_resonance)
        ):
            action = PolicyAction.ALLOW
        else:
            action = PolicyAction.RESHAPE

        rationale_parts = [
            f"verdict={verdict.value}",
            f"resonance={adjusted_resonance:.3f}",
        ]
        if divergence_status is not None:
            rationale_parts.append(f"drift={divergence_status.value}")
        if narrative_coherence is not None:
            rationale_parts.append(f"narrative={narrative_coherence:.3f}")
        rationale_parts.append(f"aggregate={aggregate_score:.3f}")
        rationale_parts.extend(
            [f"{c.name}:{c.value:.3f}x{c.weight:.2f}" for c in scores]
        )
        rationale_parts.append(f"decision={action.value}")
        return action, "; ".join(rationale_parts)

File: typescript/src/engine/policy/rule-based.ts
import type { DivergenceStatus } from "../../schema/enums.js";
import { DivergenceStatus as DS } from "../../schema/enums.js";
import type { VerdictType } from "../../schema/enums.js";
import { VerdictType as VT } from "../../schema/enums.js";
import type { PolicyAction } from "../../schema/policy.js";
import type { PolicyThresholds } from "../../schema/policy.js";
import type { ScoreComponent } from "../../schema/policy.js";

export class RuleBasedPolicyModel {
  readonly name = "rule_based";
  readonly version = "1";

  decide(params: {
    verdict: VerdictType;
    adjustedResonance: number;
    divergenceStatus: DivergenceStatus | null;
    narrativeCoherence: number | null;
    scores: ScoreComponent[];
    aggregateScore: number;
    thresholds: PolicyThresholds;
  }): { action: PolicyAction; rationale: string } {
    const { verdict, adjustedResonance, divergenceStatus, scores, aggregateScore, thresholds } =
      params;
    const driftOk =
      divergenceStatus === null ||
      divergenceStatus === DS.STABLE ||
      divergenceStatus === thresholds.maxDriftStatus;

    const effectSc = scores.find((c) => c.name === "effect_risk");
    if (
      thresholds.effectRiskMinScore !== null &&
      effectSc !== undefined &&
      effectSc.weight > 0 &&
      effectSc.value < thresholds.effectRiskMinScore
    ) {
      return {
        action: "DENY",
        rationale: `verdict=${verdict}; effect_risk=${effectSc.value.toFixed(3)} below min ${thresholds.effectRiskMinScore.toFixed(3)}; ${effectSc.explanation}`,
      };
    }

    let action: PolicyAction;
    if (verdict === VT.INCOHERENT || divergenceStatus === DS.DIVERGED) {
      action = "DENY";
    } else if (
      verdict === VT.COHERENT &&
      adjustedResonance >= thresholds.allowMinResonance &&
      driftOk &&
      aggregateScore >= thresholds.allowMinResonance
    ) {
      action = "ALLOW";
    } else {
      action = "RESHAPE";
    }

    const parts = [
      `verdict=${verdict}`,
      `resonance=${adjustedResonance.toFixed(3)}`,
    ];
    if (divergenceStatus !== null) {
      parts.push(`drift=${divergenceStatus}`);
    }
    if (params.narrativeCoherence !== null && params.narrativeCoherence !== undefined) {
      parts.push(`narrative=${params.narrativeCoherence.toFixed(3)}`);
    }
    parts.push(`aggregate=${aggregateScore.toFixed(3)}`);
    parts.push(...scores.map((c) => `${c.name}:${c.value.toFixed(3)}x${c.weight.toFixed(2)}`));
    parts.push(`decision=${action}`);
    return { action, rationale: parts.join("; ") };
  }
}
